classes/local/tag/tag_manager.php

```diff
@@ -0,0 +1,183 @@
+
+/**
+ * Manages object tagging feature.
+ *
+ * @package   tool_objectfs
+ * @author    Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class tag_manager {
+
+    /**
+     * @var int If object needs sync. These will periodically be picked up by scheduled tasks and queued for syncing.
+     */
+    public const SYNC_STATUS_NEEDS_SYNC = 0;
+
+    /**
+     * @var int Object does not need sync. Will be essentially ignored in tagging process.
+     */
+    public const SYNC_STATUS_SYNC_NOT_REQUIRED = 1;
+
+    /**
+     * @var int Object tried to sync but there was an error. Will make it ignored and must be corrected manually.
+     */
+    public const SYNC_STATUS_ERROR = 2;
+
+    /**
+     * @var array All possible tag sync statuses.
+     */
+    public const SYNC_STATUSES = [
+        self::SYNC_STATUS_NEEDS_SYNC,
+        self::SYNC_STATUS_SYNC_NOT_REQUIRED,
+        self::SYNC_STATUS_ERROR,
+    ];
+
+    /**
+     * Returns an array of tag_source instances that are currently defined.
+     * @return array
+     */
+    public static function get_defined_tag_sources(): array {
+        // All possible tag sources should be defined here.
+        // Note this should be a maximum of 10 sources, as this is an AWS limit.
+        return [
+            new mime_type_source(),
+            new environment_source(),
+        ];
+    }
+
+    /**
+     * Is the tagging feature enabled and supported by the configured fs?
+     * @return bool
+     */
+    public static function is_tagging_enabled_and_supported(): bool {
+        $enabledinconfig = !empty(get_config('tool_objectfs', 'taggingenabled'));
+
+        $client = manager::get_client(manager::get_objectfs_config());
+        $supportedbyfs = !empty($client) && $client->supports_object_tagging();
+
+        return $enabledinconfig && $supportedbyfs;
+    }
+
+    /**
+     * Gathers the tag values for a given content hash
+     * @param string $contenthash
+     * @return array array of key=>value pairs, the tags for the given file.
+     */
+    public static function gather_object_tags_for_upload(string $contenthash): array {
+        $tags = [];
+        foreach (self::get_defined_tag_sources() as $source) {
+            $val = $source->get_value_for_contenthash($contenthash);
+
+            // Null means not set for this object.
+            if (is_null($val)) {
+                continue;
+            }
+
+            $tags[$source->get_identifier()] = $val;
+        }
+        return $tags;
+    }
+
+    /**
+     * Stores tag records for contenthash locally
+     * @param string $contenthash
+     * @param array $tags
+     */
+    public static function store_tags_locally(string $contenthash, array $tags) {
+        global $DB;
+
+        // Purge any existing tags for this object.
+        $DB->delete_records('tool_objectfs_object_tags', ['contenthash' => $contenthash]);
```

Comment:
No this was purely for simplicity. Its so much easier and performant to just do a bulk delete records and a bulk insert, as opposed to an update diff.

A diff also has problems if you ever change the key of a tag, it would unlink it and the old records would float around and need to be handled, as opposed to doing it this way which ensures the old state is always wiped away and replaced with the new state.